cpp/v2/test/api_test_v2.cc

```diff
@@ -0,0 +1,262 @@
+            expected_func);
+
+  auto expected_func1 = RayMemberFunc(&Base::overload_func, cv_none, int, int);
+  EXPECT_EQ(get_function(RayMemberFunc(&Base::overload_func, cv_none, int, int)),
+            expected_func1);
+}
+
+TEST(RayApiTestV2, CallFunction) {
+  auto f = get_function(hello);
+  call_func(f);
+
+  auto f1 = get_function(bar);
+  EXPECT_EQ(call_func(f1, 1), 1);
+
+  auto f2 = get_function(RayFunc(overload_func, int));
+  EXPECT_EQ(call_func(f2, 1), 1);
+
+  auto f3 = get_function(RayFunc(overload_func));
+  EXPECT_EQ(call_func(f3), 0);
+
+  auto f4 = get_function(RayFunc(overload_func, int, int));
+  // The function was not registered, call_func with nullptr will throw
+  // std::invalid_argument.
+  EXPECT_THROW(call_func(f4, 1, 2), std::invalid_argument);
+
+  Base base{};
+  auto f5 = get_function(&Base::bar);
+  EXPECT_EQ(call_func(base, f5, 1), 1);
+
+  auto f6 = get_function(RayMemberFunc(&Base::overload_func, cv_none, int));
+  EXPECT_EQ(call_func(base, f6, 1), 1);
+
+  auto f7 = get_function(RayMemberFunc(&Base::overload_func, cv_none, int, int, int));
+  EXPECT_THROW(call_func(base, f7, 1, 2, 3), std::invalid_argument);
+}
+
+TEST(RayApiTestV2, Init) {
+  // Default arguments.
+  EXPECT_TRUE(ray::Init(num_cpus = 8));
+  EXPECT_TRUE(ray::Init(num_cpus = 8, object_store_memory = 400));
+
+  // No order arguments.
+  EXPECT_TRUE(ray::Init(object_store_memory = 400, num_gpus = 8));
+  EXPECT_TRUE(ray::Init(num_cpus = 8, ray_address = "auto"));
+}
+
+TEST(RayApiTestV2, NormalTask) {
+  auto obj = ray::Task(hello).Remote();
+  obj.Get();
+
+  auto obj1 = ray::Task(bar).Remote(1);
+  EXPECT_EQ(obj1.Get(), 1);
+
+  EXPECT_THROW(ray::Task(not_registered_func).Remote(), std::invalid_argument);
+
+  auto obj2 = ray::Task(RayFunc(overload_func, int)).Remote(1);
+  EXPECT_EQ(obj2.Get(), 1);
+
+  EXPECT_EQ(ray::Task(add).Remote(2, 3).Get(), 5);
+  EXPECT_EQ(ray::Task(add).Remote(ray::Put(1), 3).Get(), 3);
+  EXPECT_EQ(ray::Task(add).Remote(2, ray::Put(1)).Get(), 2);
+  EXPECT_EQ(ray::Task(add).Remote(ray::Put(1), ray::Put(2)).Get(), 0);
+}
+
+TEST(RayApiTestV2, NamespaceNormalTask) {
+  auto obj1 = ray::Task(bar).Remote(2);
+  EXPECT_EQ(obj1.Get(), 2);
+
+  auto obj2 = ray::Task(my_namespace::bar).Remote(2);
+  EXPECT_EQ(obj2.Get(), 3);
+
+  EXPECT_EQ(ray::Task(bar).Remote(2).Get(), 2);
+  EXPECT_EQ(ray::Task(my_namespace::bar).Remote(2).Get(), 3);
+}
+
+TEST(RayApiTestV2, OverlaodNormalTask) {
+  auto obj = ray::Task(RayFunc(overload_func)).Remote();
+  EXPECT_EQ(obj.Get(), 0);
+
+  auto obj1 = ray::Task(RayFunc(overload_func, int)).Remote(2);
+  EXPECT_EQ(obj1.Get(), 2);
+}
+
+TEST(RayApiTestV2, CreateActor) {
+  auto actor_handle = ray::Actor<Base>().Remote();
+  EXPECT_EQ(typeid(actor_handle), typeid(ray::ActorHandle<Base>));
+
+  auto actor_handle2 = ray::Actor<Base>().Remote(1);
+  EXPECT_EQ(typeid(actor_handle2), typeid(ray::ActorHandle<Base>));
+
+  auto actor_handle3 = ray::Actor<Base>().Remote(1, ObjectRef<int>{});
```

Comment:
what is this?

Reply:
This case to show passing int or ObjectRef<int> both are ok, ray api will do arguments boxing and unboxing automatically. It's free and flexible for user.